Amqp/src/Akka.Streams.Amqp/Model.cs

```diff
@@ -0,0 +1,615 @@
+                ConnectionTimeout, HandshakeTimeout, NetworkRecoveryInterval, AutomaticRecoveryEnabled,
+                topologyRecoveryEnabled);
+        }
+
+
+        public override string ToString()
+        {
+            return
+                $"AmqpConnectionDetails(HostAndPortList=({HostAndPortList.Select(x => $"[{x.host}:{x.port}]").Aggregate((left, right) => $"{right}, {left}")}), Credentials={Credentials}, VirtualHost={VirtualHost})";
+        }
+    }
+
+    public struct AmqpCredentials : IEquatable<AmqpCredentials>
+    {
+        public string Username { get; }
+        public string Password { get; }
+
+        private AmqpCredentials(string username, string password)
+        {
+            Username = username;
+            Password = password;
+        }
+
+        public static AmqpCredentials Create(string username, string password)
+        {
+            return new AmqpCredentials(username, password);
+        }
+
+        public override string ToString()
+        {
+            return $"AmqpCredentials(Username={Username}, Password=********)";
+        }
+
+        public override bool Equals(object obj)
+        {
+            return obj is AmqpCredentials && Equals((AmqpCredentials)obj);
+        }
+
+        public bool Equals(AmqpCredentials other)
+        {
+            return Username == other.Username &&
+                   Password == other.Password;
+        }
+
+        public override int GetHashCode()
+        {
+            var hashCode = 568732665;
+            hashCode = hashCode * -1521134295 + base.GetHashCode();
+            hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(Username);
+            hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(Password);
+            return hashCode;
+        }
+
+        public static bool operator ==(AmqpCredentials credentials1, AmqpCredentials credentials2)
+        {
+            return credentials1.Equals(credentials2);
+        }
+
+        public static bool operator !=(AmqpCredentials credentials1, AmqpCredentials credentials2)
+        {
+            return !(credentials1 == credentials2);
+        }
+    }
+
+    public interface IDeclaration
+    {
+    }
+
+    public sealed class QueueDeclaration : IDeclaration
+    {
+
+        private QueueDeclaration(string name, bool durable = false, bool exclusive = false, bool autoDelete = false,
+            IDictionary<string, object> arguments = null)
+        {
+            Name = name;
+            Durable = durable;
+            Exclusive = exclusive;
+            AutoDelete = autoDelete;
+            Arguments = arguments ?? new Dictionary<string, object>();
+        }
+
+        public string Name { get; }
+
+        public bool Durable { get; }
+
+        public bool Exclusive { get; }
+
+        public bool AutoDelete { get; }
+
+        public IDictionary<string, object> Arguments { get; }
```

Comment:
If essentially the whole type is immutable and all mutators are pure - use immutable collection too.